### ISSUE TITLE
NLog V4.4.2 abruptly stops logging for no apparent reason

### ISSUE THREAD

**Issue by ReeceWilliams** [NLog/NLog#1998]
Bug
**NLog version**:  4.4.2 and later
**Platform**:  .Net 4.5
**Current NLog config**
We have a C# batch process that connects to several different customer specific databases and we create a log file per database at start up.
We programmatically create log file targets based on the databases that the process connects to. The resultant log file instances are cached in a static class. The code we use is below.

This has all been working fine until (by a process of elimination) we went later than V4.4.1 - so 4.4.3 exhibits this as well, V4.4.1 works ok.

```c#
public static bool AddDatabaseTarget(string dbName) {
	if (File.ContainsKey(dbName)) return false;

	var config = LogManager.Configuration;

	// allow 10 database specific log files before rollover
	AddTarget(dbName, config, LogLevel.Info, 10);
	LogManager.Configuration = config;

	var logFile = new LogFile(dbName);
	if (logFile.LogFileTargetValid) File.TryAdd(dbName, logFile);
	return true;
}

private static void AddTarget(string dbName, LoggingConfiguration config, LogLevel level, int? maxArchiveFilesOverride = null)
{
	var atw = new AsyncTargetWrapper
	{
		Name = dbName,
		WrappedTarget = new FileTarget
		{
			Name = dbName,
			FileName = Path.Combine(RootLogFolder, dbName, dbName + ".log"),
			ArchiveFileName = Path.Combine(RootLogFolder, dbName, dbName + ".{#}.log"),
			MaxArchiveFiles = maxArchiveFilesOverride ?? MaxArchiveFiles,
			ArchiveAboveSize = ArchiveAboveSize,
			ArchiveNumbering = ArchiveNumberingMode.Sequence,
			ArchiveEvery = FileArchivePeriod.Day,
			KeepFileOpen = false,
			CreateDirs = true,
			Layout = "${longdate} ${uppercase:${level}}\t${message}"
		}
	};
	LogManager.Configuration.AddTarget(dbName, atw);
	config.LoggingRules.Add(new LoggingRule(dbName, level, atw) { Final = true });
}
```
The effect is that logging just stops; it happily logs things out until a certain point when logging ceases.

The internal, info level log files are identical between V4.4.1 (which is working) and V4.4.2 (which stops logging):
```
2017-03-02 10:10:06.0238 Info Configured from an XML element in D:\Batch\Alerts\AlertsGeneration.exe.Config...
2017-03-02 10:10:06.2737 Info Found 0 configuration items
2017-03-02 10:10:06.3832 Info Configuration initialized.
2017-03-02 10:10:06.3987 Info NLog, Version=4.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c. File version: 4.4.4471. Product version: 4.4.2.
2017-03-02 10:10:07.1019 Info Closing old configuration.
2017-03-02 10:10:07.1175 Info Found 75 configuration items
2017-03-02 10:10:07.1175 Info Found 75 configuration items
2017-03-02 10:10:08.4456 Info Closing old configuration.
2017-03-02 10:10:08.4456 Info Found 90 configuration items
2017-03-02 10:10:08.4456 Info Found 90 configuration items
2017-03-02 10:10:10.5550 Info Closing old configuration.

<snip />

2017-03-02 10:10:12.8988 Info Found 465 configuration items
2017-03-02 10:10:12.9457 Info Found 465 configuration items
2017-03-02 10:10:13.1957 Info Shutting down logging...
2017-03-02 10:10:13.1957 Info Logger has been shut down.
```

Has something changed that is inadvertently limiting the number of log files you can generate at a time?

I'm pretty sure we're up to 30 separate log files now.
Thanks.

**Comment by 304NotModified** [comment]
Is it possible to create a trace internal log?

@snakefoot any idea on this?

**Comment by ReeceWilliams** [comment]
[TraceLogCleaned.zip](https://github.com/NLog/NLog/files/812634/TraceLogCleaned.zip)
Trace log attached - I had to obfuscate the actual database names for client confidentiality.

I have also tried different batch projects and they've exhibited the same behaviour of working ok with 4.4.1, stopping with 4.4.2.

We also use the same logging set up with our web api, but that's not so easy to experiment with; I'm assuming it will behave the same...

On a side note, I'm adding one target at a time, because I think I originally had issues adding more than that in one go - ie doing:
- Get config
- Add target
- Set Config
as opposed to:
 - Get config
- Add all of the targets at once 
- Set Config

It's a while since I've looked at this code so my memory's a little hazy, but should the all-at-once thing work?

Thanks.

**Comment by ReeceWilliams** [comment]
...also, this is the timing of the last log entry in the batch log file, compared with the entries in the trace log:
```
2017-03-02 12:33:22.5765 Trace AsyncWrapper 'Alerts': writing 2 events (Timer)
2017-03-02 12:33:22.5765 Trace Opening D:/Logs/Batch\Alerts\Alerts.log with allowFileSharedWriting=False

 -- 2017-03-02 12:33:22.6077  This is the timestamp of the last entry in the batch log --

2017-03-02 12:33:22.7327 Trace AsyncWrapper 'Alerts': writing 2 events (Timer)
2017-03-02 12:33:22.7327 Trace Opening D:/Logs/Batch\Alerts\Alerts.log with allowFileSharedWriting=False
2017-03-02 12:33:22.7640 Info Shutting down logging...
2017-03-02 12:33:22.7640 Info Logger has been shut down.
```

**Comment by ReeceWilliams** [comment]
...and there's more...

This is the timestamp of the first log entry in the batch log, compared with the entries in the trace log:
```
2017-03-02 12:33:09.1389 Trace Initializing AsyncWrapper Target[SMSBatch](File Target[SMSBatch])
2017-03-02 12:33:09.1389 Trace Initializing Layout Renderer: ${message}
2017-03-02 12:33:09.1389 Trace Initializing Layout Renderer: ${literal}

-- 2017-03-02 12:33:09.1545 -- This is the timestamp of the first log entry in the Alerts target --

2017-03-02 12:33:09.1545 Trace Initializing Layout Renderer: ${level}
2017-03-02 12:33:09.1545 Trace Initializing '${level}'
2017-03-02 12:33:09.1545 Trace Initializing Layout Renderer: ${uppercase}
```

**Comment by snakefoot** [comment]
@304NotModified The cause is probably the "constant" changing of LogManager.Configuration.

Maybe NLog forgets to ignore (or block) the loggers while the NLog-configuration is in progress of being closed/initialized?

**Comment by 304NotModified** [comment]
@snakefoot that sounds plausible

**Comment by snakefoot** [comment]
> It's a while since I've looked at this code so my memory's a little hazy, but should the all-at-once thing work?

@ReeceWilliams Yes it should work. Re-assigning the Configuration for each target added, actually generates a huge overhead as all existing targets are opened/closed on each configuration-assignment.

Do you have one thread that adds one target at a time the NLog-Configuration, and other threads that does the actual logging while the targets are added? (Curious how we can make a small test application to reproduce the bug)

**Comment by ReeceWilliams** [comment]
I know it didn't feel right doing one target at a time, so I must have had a good reason to do it that way originally...I'll have to give it another go and see what happens.

I'm seeing this it in a single threaded batch process although we do use this in our web api as well - the logfile dictionary I add to is a Lazy<T> which should be thread safe in any case.

First thing I do is create a single batch target to log progress to.
Next it gets a list of databases it can work with, which causes the DB specific targets to be created and then queries each DB to see if anything needs to be processed, logging the details to the batch target.

Looking at the log output, it looks like the DB targets are still being created after the process has moved on to checking whether there's anything needs processing, so I guess target creation is asynchronous?

I guess there could be a problem if I'm trying to write to the batch log while in the middle of reconfiguring a long list of targets.

I'll look at refactoring the target creation and see how I go.
Thanks.

**Comment by snakefoot** [comment]
> I guess target creation is asynchronous?

Nope, when the LogManger.Configuration-setter has completed, then all NLog-targets initialized.

Think you need to give more detail of how you do this:

1. Gets a list of databases
2. Iterates list of databases
   2.1 Adds file-logger-target for database (LogManager.Configuration assigned)
   2.2 Schedules database query for stuff to process (Does this happen on a new thread?)
   2.3 Processes the result of the database-query (Does this happen on a new thread?)

**Comment by ReeceWilliams** [comment]
Sorry for the delay in responding - I had to pull this all together.

I've refactored the code to reduce the number of re-config calls; it's now down to three.
I'm still seeing the log messages stopping.

Attached zip file contains:
1) The alerts.log, showing where logging is stopping.
2) Log.cs: This is the class used to create targets and it stores a dictionary of the targets defined.
3) Program.cs: This is an excerpt from the batch process entry point, showing where the log messages are coming from
4) TraceLog.txt: The nlog trace log

**Further Explanation:**

The batch process starts and logs to the batch log by calling Log.Batch() (Program.L5).

This references the Log.File dictionary and causes the Lazy<T> (LazyFile) lamda to trigger (Log.L203), creating the "fixed" log files - these are the ones we always create and are commonly used in the web api, but not so much in the batch process.

We then get a list of databases (Program.L41) by calling a different database cache class "My" (excerpt below), which works in a similar manner.
Referencing "My.Db" causes the Lazy<T> lambda to trigger (outlined below) .

This lambda first adds the "Master" database target, followed by each of the other customer database targets.

I've refactored things as previously discussed, so that all the customer database logs are added in a single re-config, rather than individually, however Master contains the list of customer databases to configure, so the target for master has to be defined before the remaining customer database targets, just so we can log messages there if there are issues accessing Master. 
Conceivably, I can add this to the list of "well known" targets that we created earlier and avoid an additional re-config.
 
The "My" database class is shaped like this:
```
public class My
{
    public static string MasterDbName { get; } = "Master"
		
	private static readonly Lazy<ReadOnlyDictionary<string, SqlData>> LazyDb = new Lazy<ReadOnlyDictionary<string, SqlData>>(() =>
	{
		...
               Log.AddDatabaseTarget(MasterDbName);
		...
		
		// get a list of potential databases to connect to
				
		...
		Log.AddDatabaseTargets([string array of DB names])
		...
	}

	public static ReadOnlyDictionary<string, SqlData> Db => LazyDb.Value;
}
```

Any multi-threading is happening within the Lazy<T> lambda's which are guaranteed to be threadsafe.

I hope that all makes sense - please let me know if you have further questions.
[CodeAndLogs.zip](https://github.com/NLog/NLog/files/826313/CodeAndLogs.zip)

Thanks,
Reece.

**Comment by snakefoot** [comment]
@ReeceWilliams It is a little confusing of source-code-files, log-files and pseudo-code. Would be easier with an isolated console-application.

When looking in Alerts.log, then the last entry is this:

2017-03-08 10:28:40.6602 INFO	Processing Database21

And when looking in Tracelog.txt, then the last entries are these:

2017-03-08 10:28:40.6133 Trace AsyncWrapper 'Alerts': writing 10 events (Timer)
2017-03-08 10:28:40.6133 Trace Opening D:/Logs/Batch\Alerts\Alerts.log with allowFileSharedWriting=False
2017-03-08 10:28:40.6758 Trace AsyncWrapper 'Alerts': writing 10 events (Timer)
2017-03-08 10:28:40.6758 Trace Opening D:/Logs/Batch\Alerts\Alerts.log with allowFileSharedWriting=False
2017-03-08 10:28:40.7071 Info Shutting down logging...
2017-03-08 10:28:40.7071 Info Logger has been shut down.

So 40 milliseconds after the last log-entry in the Alerts.log, then NLog reacts to an application-shutdown-event. I'm not guessing it was intended to have this application-shutdown.

Can you check your Windows EventLog (Application / System) to see if your application or NET-Runtime have written anything for this premature application-shutdown?

If nothing is written in the EventLog could you add an UnhandledExceptionHandler and dump the details of this exception. Something like this:

```
AppDomain.CurrentDomain.UnhandledException += new UnhandledExceptionEventHandler(CurrentDomain_UnhandledException);

static void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
{
    Exception ex = e.ExceptionObject as Exception;
    using (var textWriter = File.AppendText(@"D:\Logs\Batch\Alerts\UnhandledException.log"))
    {
        textWriter.WriteLine(ex != null ? ex.ToString() : "Unknown Exception);
    }
}
```

**Comment by snakefoot** [comment]
@ReeceWilliams Also I think it is bad idea to give both the AsyncTargetWrapper + FileTarget the same name. Instead it should be AsyncTargetWrapper.Name = dbName and FileTarget.Name = dbName + "Wrapped".

**Comment by ReeceWilliams** [comment]
The shutdown is the batch process completing; it trawls through all of the databases looking for things to process.  In this case there's nothing to process and it shuts down.

Whilst the logging stops, the process does complete successfully.

**Comment by snakefoot** [comment]
@ReeceWilliams I'm confused. To investigate this bug, then you have provided log files, where everything is working?

Anyways it is a good idea (for any application log), to write when the application has completed, and is about to shutdown.

But I'm kind of tired of this now. So until you have provided a console-application, that can atleast re-produce the case 1 out of 4 times being run, then I will not do anything further.

**Comment by snakefoot** [comment]
@ReeceWilliams My mind have suddenly become curious of why the shutdown is not performing a target flush operation. Will try and investigate this over the weekend.

**Comment by snakefoot** [comment]
Think the bug is here (sender-parameter is not a LogFactory):

```
        private static void OnStopLogging(object sender, EventArgs args)
        {
            try
            {
                var logFactory = sender as LogFactory;
                if (logFactory != null)
                {
                    // Finalizer thread has about 2 secs, before being terminated
                    logFactory.Close(TimeSpan.FromMilliseconds(1500));
                }
```

So it never flushes  and closes properly on process exit.

**Comment by 304NotModified** [comment]
Fixed in NLog 4.4.5 thanks to @snakefoot  !

**Comment by snakefoot** [comment]
@ReeceWilliams Please check if NLog 4.4.5 resolves your issue: https://github.com/NLog/NLog/releases/tag/v4.4.5-beta1
